src/ResizeConsumer.tsx

```diff
@@ -1,69 +1,90 @@
-import * as PropTypes from 'prop-types';
 import * as React from 'react';
 import { omit } from 'lodash';
 
-import { HTMLAttributes, RefObject } from 'react';
+import { HTMLAttributes, ReactNode, RefObject, SFC } from 'react';
 
 import { Consumer, ContextValue } from './context';
 import Listener from './Listener';
-import { isRefObject, updateElementDataAttributes } from './methods';
+import { updateElementDataAttributes } from './methods';
 
 import { Size } from './types';
 
-interface ForwardRefProps extends HTMLAttributes<HTMLDivElement> {
+interface ExternalProps extends HTMLAttributes<HTMLDivElement> {
+  innerRef?: RefObject<HTMLDivElement>;
   onSizeChanged?: (size: Size) => void;
   updateDatasetBySize?: (size: Size) => DOMStringMap;
+  children?: ReactNode | SFC<Size>;
 }
 
-interface Props extends ForwardRefProps {
+interface Props extends ExternalProps {
+  innerRef: RefObject<HTMLDivElement>;
   context: ContextValue;
-  forwardedRef: RefObject<HTMLDivElement>;
 }
 
-class ResizeConsumer extends React.PureComponent<Props> {
+interface State {
+  children: ReactNode;
+}
+
+class ResizeConsumer extends React.PureComponent<Props, State> {
+  state: State = {
+    children: null,
+  };
+
   private currentListenElement: HTMLElement | null = null;
 
   private get divProps() {
     return omit(this.props, [
-      'context',
-      'forwardedRef',
+      'innerRef',
       'onSizeChanged',
       'updateDatasetBySize',
+      'children',
+      'context',
     ]);
   }
 
-  public componentDidMount() {
+  private get children() {
+    return typeof this.props.children === 'function' ? this.state.children : this.props.children
+  }
+
+  componentDidMount() {
     this.updateListener();
   }
 
-  public componentDidUpdate() {
+  componentDidUpdate() {
     this.updateListener();
   }
 
-  public componentWillUnmount() {
+  componentWillUnmount() {
     this.stopListen();
   }
 
-  public render() {
+  render() {
     return (
-      <div {...this.divProps} ref={this.props.forwardedRef}>
-        {this.props.children}
+      <div {...this.divProps} ref={this.props.innerRef}>
+        {this.children}
       </div>
     );
   }
 
   private updateAttribute = (size: Size) => {
-    const element = this.props.forwardedRef.current;
+    const element = this.props.innerRef.current;
     if (element && typeof this.props.updateDatasetBySize === 'function') {
       const newDataAttributes = this.props.updateDatasetBySize(size);
       updateElementDataAttributes(element, newDataAttributes);
     }
   };
 
   private onSizeChanged = (size: Size) => {
+    if (typeof this.props.children === 'function') {
+      this.setState({
+        children: this.props.children(size)
+      });
```

Comment:
I prefer to store size into state here, and implement `render()` with:
```jsx
render() {
    return (
      <div {...this.divProps} ref={this.props.innerRef}>
        {typeof this.props.children === 'function'
          ? this.props.children(this.state.size)
          : this.props.children}
      </div>
    );
 }
```
It's not quite good to store a ReactNode into state IMO. What do you think about it?